src/main/java/org/openrewrite/staticanalysis/ReplaceWeekYearWithYear.java

```diff
@@ -0,0 +1,120 @@
+               " December and first week of January Week Year could produce unexpected results.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("RSPEC-3986");
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(
+                Preconditions.or(
+                        new UsesType<>("java.util.Date", false),
+                        new UsesType<>("java.time.format.DateTimeFormatter", false),
+                        new UsesType<>("java.text.SimpleDateFormat", false)
+                ),
+                new ReplaceWeekYearVisitor()
+        );
+    }
+
+    private static class ReplaceWeekYearVisitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
+            if (OF_PATTERN_MATCHER.matches(mi)) {
+                getCursor().putMessage("KEY", mi);
+            }
+
+            return super.visitMethodInvocation(mi, ctx);
+        }
+
+        @Override
+        public J.NewClass visitNewClass(J.NewClass nc, ExecutionContext ctx) {
+            if (SIMPLE_DATE_FORMAT_CONSTRUCTOR_MATCHER.matches(nc)) {
+                getCursor().putMessage("KEY", nc);
+            }
+
+            return super.visitNewClass(nc, ctx);
+        }
+
+        @Override
+        public J.Literal visitLiteral(J.Literal li, ExecutionContext ctx) {
+            if (li.getValue() instanceof String) {
+                Cursor c = getCursor().dropParentWhile(is -> is instanceof J.Parentheses || !(is instanceof Tree));
+                if (c.getMessage("KEY") != null) {
+                    String value = li.getValueSource();
+
+                    if (value == null) {
+                        return li;
+                    }
+
+                    String newValue = replaceY(value);
+
+                    return li.withValueSource(newValue).withValue(newValue);
+                }
+            }
+
+            return li;
+        }
+
+        public static String replaceY(String input) {
+            StringBuilder output = new StringBuilder();
+            boolean insideQuotes = false;
+
+            for (int i = 0; i < input.length(); i++) {
+                char currentChar = input.charAt(i);
+                char nextChar = (i < input.length() - 1) ? input.charAt(i + 1) : '\0';
+
+                if (currentChar == '\'') {
+                    insideQuotes = !insideQuotes;
+                    output.append(currentChar);
+                } else if (currentChar == 'Y' && !insideQuotes) {
+                    output.append('y');
+                } else if (currentChar == 'Y' && nextChar == '\'') {
+                    output.append(currentChar);
+                } else {
+                    output.append(currentChar);
+                }
+            }
+
+            return output.toString();
```

Comment:
I suspect that when we do the `J.Literal#withValue()` call it may end up producing a new `J.Literal` object. So I think we should here in the loop update a `boolean` variable if there was any change made and only then return this result. Otherwise the input value to this method. Does that make sense to you?

Reply:
Ah I see, ya I can add a check to see if there are any changes actually being made.